### DISCUSSION TITLE
Sorted map output without BTF

### DISCUSSION THREAD

**Discussion by mtijanic** [bpftrace/bpftrace#3302] [Q&A]
I have a custom map, populated like `@map[$key] = ($a, $b, $c);`, all int variables. Then, I want to do something like...
```
END {
    for ($kv : @map) {
        // Custom printing format, simplified for the sake of the example
        printf("%x = %x | %x | %x\n", $kv.0, $kv.1.0, $kv.1.1, $kv.1.2);
     }
}
```
but the order in which it is iterated is effectively entirely random. Let's say I wanted it sorted by $a. I figured I could do something like this...

```
probe {
    @map[$key] = ($a, $b, $c);
    @for_sorting = max($a);
}
END {
    for ($kv : @for_sorting) {
        // Custom printing format, simplified for the sake of the example
        printf("%x = %x | %x | %x\n", $kv.0, @map[$kv.0].0, @map[$kv.0].1, @map[$kv.0].2);
     }
}
```
and rely on implicit sorting when printing builtin maps. But, you can't iterate any of the builtin maps without BTF, and I'd like to deploy this on a wide range of machines so can't assume it's present.

Is there any trick to get arbitrary sorting? Implement some sort of insertion sort prior to printing?

**Comment by mtijanic** [comment]
Stupid-but-it-works O(n^2) solution:
```
    @last_max_val = 0xffffffff;
    for ($tmp : @map) {
        @max_val = 0;
        @max_key = 0;
        for ($kv : @map) {
            if (($kv.1.0 > @max_val) && ($kv.1.0 < @last_max_val) && ($kv.0 != @max_key)) {
                @max_val = $kv.1.0;
                @max_key = $kv.0;
            }
        }
        if (@max_key != @last_max_key) {
            $a = @map[@max_key].0;
            $b = @map[@max_key].1;
            $c = @map[@max_key].2;

            printf("%x = %x | %x | %x\n", @max_key, $a, $b, $c);
            @last_max_val = @max_val;
            @last_max_key = @max_key;
        }
    }
    delete(@max_key);
    delete(@max_val);
    delete(@last_max_key);
    delete(@last_max_val);
```
would still love to hear if there's a better one.

**Comment by ajor** [comment]
You inspired me to have a go myself! Here's my solution using heapsort, which should theoretically be O(n log(n)). It's using a hashmap (`@buf`) as its scratch space since bpftrace doesn't provide array storage.

It sorts based on the first value in the map (`$kv.1.0`):
```
BEGIN
{
  @[1] = (10, 100);
  @[2] = (20, 200);
  @[3] = (30, 300);
  @[4] = (40, 400);
  @[5] = (50, 500);


  $i = 0;
  for ($kv : @) {
    @buf[$i++] = $kv;
  }

  // Heapsort
  $count = len(@);

  // !!! Have to hardcode in $start and $end or it doesn't work!
  $start = 2; //$count / 2;
  $end = 5; //$count;
  while ($end > 1) {
    if ($start > 0) {
      $start -= 1;
    } else {
      $end -= 1;

      $tmp = @buf[$end];
      @buf[$end] = @buf[0];
      @buf[0] = $tmp;
    }

    $root = $start;
    // iLeftChild(i) = 2i + 1;
    while (2*$root + 1 < $end) {
      $child = 2*$root + 1;
      if ($child + 1 < $end && @buf[$child].1.0 < @buf[$child+1].1.0) {
        $child = $child + 1;
      }

      if (@buf[$root].1.0 < @buf[$child].1.0) {
        $tmp = @buf[$root];
        @buf[$root] = @buf[$child];
        @buf[$child] = $tmp;
        $root = $child;
      } else {
        break;
      }
    }
  }

  // Print the sorted buffer
  $i = 0;
  while ($i < $count) {
    $kv2 = @buf[$i++];
    printf("%d %d %d\n", $kv2.0, $kv2.1.0, $kv2.1.1);
  }

  clear(@buf);
  clear(@);
  exit();
}
```

I think I've found a few bpftrace bugs working on this :)
1. I have to hardcode the values 2 and 5 for `$start` and `$end` or the loop conditions don't work correctly
2. The maps in your solution above (e.g. `@last_max_val`, `@max_val`) should be convertible to local variables (`$last_max_val`, `$max_val`) in the latest (unreleased) bpftrace, but that causes LLVM assertions / segfaults!

**Reply by ajor** [reply]
I've opened issues for the bugs found here: #3307 #3308
